If compiling against libc++ without filesystem, don't reference fstream

stevew817 opened the pull request:
SetupPayloadHelper.cpp is compiled as part of the `setup_payload` static library, which is built for (almost?) all targets, even if those targets may not have an onboard filesystem. For GCC's libstdc++ this doesn't matter as it still provides a stub for std::ifstream, but LLVM's libc++ doesn't and throws a compiletime error because of this reference.

The alternative to this patch is to move out SetupPayloadHelper.cpp from the `setup_payload` library into the `qrcodetool` executable, since that seems to be the only place the contents of SetupPayloadHelper are in use?


#### Testing

Tested on EFR32 with GCC and ARM LLVM toolchains.

andy31415 commented on src/setup_payload/SetupPayloadHelper.cpp:
```diff
@@ -143,6 +143,11 @@ static CHIP_ERROR addParameter(SetupPayload & setupPayload, const SetupPayloadPa
 
 CHIP_ERROR loadPayloadFromFile(SetupPayload & setupPayload, const std::string & filePath)
 {
+#if defined(_LIBCPP_HAS_FILESYSTEM) && _LIBCPP_HAS_FILESYSTEM == 0
```
This seems to stub things out at the lowest level.

How about instead just removing the `SetupPayloadHelper` from the build system instead? As it seems impossible to `generateQRCodeFromFilePath` and `generateManualCodeFromFilePath` when there is no file system ability.

Stubbing here seems to make the methods globally available and just not work.

stevew817 replied:
> Stubbing here seems to make the methods globally available and just not work.

That's what the project has been doing the whole time, and I didn't know if there was a reason for that or not, so I went for the least invasive patch. I can rework it to move this file out, but should the file itself then also be moved?

andy31415 replied:
Probably different source set at least, so people can opt in to use it. I am unsure about the blast radius ... hoping not large and only used when a filesystem actually exists.

andy31415 reviewed (changes requested):
It seems to me we should completely remove the SetupPayloadHelper from the compile units in these cases (and probably static assert we have a file system in the header to prevent bad includes).

This is more work, but it also seems much more correct.

stevew817 committed:
Split SetupPayloadHelper out of setup_payload library

SetupPayloadHelper uses filesystem calls, which may not be implemented when
using a standard library which does't provide such functionality (such as
libc++ on embedded targets). This change avoids compiling in the helper
functions into the library on these targets by default, as the only in-tree
consumer of these functions currently is the qrcodetool which is only built
on host.

stevew817 commented:
@andy31415 Does this look better to you? I didn't find a good property to `static_assert` on which isn't library-specific.
